### PR TITLE
 Allow input of more than 12 numbers and enable entry of decimal values in fast-number-field

### PR DESCRIPTION
**Pull request by saicharan0124**
# Pull Request

## 📖 Description

- Users are  not able to enter a decimal value.

To reproduce this issue :
1. Go to Number field in Storybook
2. Set the step value to 0.01
3. Try to enter 5.5

- Users are not able more that 12 numbers 
1. Go to Number field in Storybook
2. Try typing continuous 1s
**Result :** 11111111111100000000

### 🎫 Issues

#6714 
#6713 

## 👩‍💻 Reviewer Notes

For valid  values, it ensures that the value falls within the specified range (min and max). If the value is below the minimum, it is updated to the minimum value. Similarly, if the value is above the maximum, it is updated to the maximum value. This provides more accurate handling of boundary conditions.

 In case of incrementing and decrementing the value, it follows as in the MDN docs. It ensures that floating values are rounded to match the step value's precision. This ensures consistent behavior and accurate handling of floating values.

## 📑 Test Plan
before bug fix :
[bugfix2_B4.webm](https://github.com/microsoft/fast/assets/81006148/a15c0539-f1f7-4a5d-b3a4-39eea2676ec4)
after bug fix:
[bug_fix2_After.webm](https://github.com/microsoft/fast/assets/81006148/86b6d475-9e6f-4698-beca-8c7362471b2a)

### General

- [x] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

**Review by chrisdholt — COMMENTED** [review]
Thanks for the contribution @saicharan0124 - I'll take a look at this. Can you add some tests to ensure that these scenarios are covered and don't regress?

**Comment by saicharan0124** [comment]
Yes,I will gladly add tests to ensure that the scenarios related to the code changes are covered and that regressions are prevented.

File: change/@microsoft-fast-foundation-dbd57bcd-a145-4d39-8b08-9fef5b496704.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
```
Comment (chrisdholt) [inline change/@microsoft-fast-foundation-dbd57bcd-a145-4d39-8b08-9fef5b496704.json:2]:
```suggestion
  "type": "prerelease",
```

File: change/@microsoft-fast-foundation-dbd57bcd-a145-4d39-8b08-9fef5b496704.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  "comment": "Allow input of more than 12 numbers and enable entry of decimal values in fast-number-field",
+  "packageName": "@microsoft/fast-foundation",
+  "email": "saicharan0124@gmail.com",
+  "dependentChangeType": "patch"
```
Comment (chrisdholt) [inline change/@microsoft-fast-foundation-dbd57bcd-a145-4d39-8b08-9fef5b496704.json:6]:
```suggestion
  "dependentChangeType": "prerelease"
```

**Review by chrisdholt — COMMENTED** [review]
Looks like several of the tests are failing with incorrect outputs on the CI

File: packages/web-components/fast-foundation/src/number-field/number-field.ts [hunk 3]
```diff
@@ -249,12 +249,20 @@ export class FASTNumberField extends FormAssociatedNumberField {
...
+            const numericValue: number = parseFloat(parseFloat(value).toPrecision(12));
+            const minValue: number = this.min !== undefined ? this.min : numericValue;
+            const maxValue: number = this.max !== undefined ? this.max : numericValue;
+
+            // Check if the value falls within the valid range
+            if (numericValue < minValue) {
+                validValue = minValue.toString();
+            } else if (numericValue > maxValue) {
+                validValue = maxValue.toString();
+            }
```
Comment (YonatanKra) [inline packages/web-components/fast-foundation/src/number-field/number-field.ts:265]: Your code also accepts numbers like this: '5-', which are not supposed to be valid.

You are missing an `else` in the final conditional:

```suggestion
            if (numericValue < minValue) {
                validValue = minValue.toString();
            } else if (numericValue > maxValue) {
                validValue = maxValue.toString();
            } else {
                validValue = numericValue.toString();
            }
```

File: packages/web-components/fast-foundation/src/number-field/number-field.ts [hunk 4]
```diff
@@ -267,8 +275,13 @@ export class FASTNumberField extends FormAssociatedNumberField {
      */
     public stepUp(): void {
         const value = parseFloat(this.value);
+        const decimalString = this.step.toString();
+        const decimalPlaces = decimalString.includes(".")
+            ? decimalString.split(".")[1].length
```
Comment (YonatanKra) [inline packages/web-components/fast-foundation/src/number-field/number-field.ts:280]: This hard codes `.` as the decimal point, but what about i18n? Sometimes decimal point is `,` for instance...

**Comment by saicharan0124** [comment]
I apologize for not addressing this issue earlier. I'll look into it now.

**Comment by YonatanKra** [comment]
@saicharan0124 

NP. That's what code review is for :)

I implemented everything according to the HTML spec. If you'd like to look at the implementation, it's [here](https://github.com/Vonage/vivid-3/pull/1384). We already merged it to production in our code and it works great (we are still using V2 of Fast, but it's mostly the same).

**Comment by janechu** [comment]
Closing as this PR appears to be quite stale without addressing the feedback from @chrisdholt.

**Pull request closed without merging**